packages/attestation-service/src/models/attestation.ts

```diff
@@ -28,16 +46,51 @@ export default (sequelize: Sequelize) => {
     account: DataTypes.STRING,
     identifier: DataTypes.STRING,
     issuer: DataTypes.STRING,
+    countryCode: DataTypes.STRING,
+    phoneNumber: DataTypes.STRING,
+    message: DataTypes.STRING,
+    ongoingDeliveryId: DataTypes.STRING,
+    providers: DataTypes.STRING,
+    attempt: DataTypes.INTEGER,
     status: DataTypes.STRING,
-    smsProvider: DataTypes.STRING,
+    errors: DataTypes.STRING,
   }) as AttestationStatic
 
-  model.prototype.canSendSms = function() {
-    return [
-      AttestationStatus.DISPATCHING,
-      AttestationStatus.FAILED,
-      AttestationStatus.UNABLE_TO_SERVE,
-    ].includes(this.status)
+  model.prototype.key = function(): AttestationKey {
+    return { account: this.account, identifier: this.identifier, issuer: this.issuer }
+  }
+
+  model.prototype.provider = function(): string | null {
+    const pl = this.providers.split(',')
+    return this.providers ? pl[this.attempt % pl.length] : null
+  }
+
+  model.prototype.recordError = function(error: string) {
+    const errors = this.errors ? JSON.parse(this.errors) : {}
+
+    errors[this.attempt] = {
+      provider: this.provider(),
+      error,
+    }
+    this.errors = JSON.stringify(errors)
+  }
+
+  model.prototype.failure = function(): boolean {
+    return (
+      // tslint:disable-next-line: triple-equals
+      this.status == AttestationStatus.NotSent.valueOf() ||
+      // tslint:disable-next-line: triple-equals
+      this.status == AttestationStatus.Failed.valueOf()
+    )
+  }
+
+  model.prototype.currentError = function() {
+    if (this.failure()) {
+      const errors = this.errors ? JSON.parse(this.errors) : {}
+      return errors[this.attempt]?.error ?? errors[this.attempt - 1]?.error ?? undefined
```

Comment:
why does this look at the past two errors?

Reply:
because we need to differentiate between waiting for the final attempt's delivery failure, and receiving that and not attempting to send any more. If all retries are exhausted, then the final error will be recorded on the preceding attempt.